update: third tutorial revamp (#4736)
JetBrains/kotlin-web-site: docs/topics/native/mapping-function-pointers-from-c.md

MyFun supply_fun() {
  return myFun;
}
``` 

> The `interop.def` file is enough to compile and run the application or open it in an IDE.

Author: I don't think we did, fixed it across all the tutorials

> The `interop.def` file is enough to compile and run the application or open it in an IDE.

Reviewer: I might be confusing it with something else - looks great now though thank you! 💯

   ```

As you can see, C function pointers are represented in Kotlin using `CPointer<CFunction<...>>`. The `accept_fun()` function
takes an optional function pointer as a parameter, while `supply_fun()` returns a function pointer.

`CFunction<(Int) -> Int>` describes the function signature, and `CPointer<CFunction<...>>?` represents a nullable
function pointer. There is an `invoke` operator extension function available for all `CPointer<CFunction<...>>` types,
allowing you to call function pointers as if they were regular Kotlin functions.

## Pass Kotlin function as a C function pointer

It's time to try using C functions from Kotlin code. Call the `accept_fun()` function and pass the C function pointer
to a Kotlin lambda:

```kotlin
import interop.*
import kotlinx.cinterop.staticCFunction
import kotlinx.cinterop.ExperimentalForeignApi

@OptIn(ExperimentalForeignApi::class)
fun myFun() {
    accept_fun(staticCFunction<Int, Int> { it + 1 })
}
```

This call uses the `staticCFunction {}` helper function from Kotlin/Native to wrap a Kotlin lambda function into a C

> function pointer. It only allows having unbound and non-capturing lambda functions. For example, it cannot use a local

Reviewer: ```suggestion
function pointer. It allows only unbound and non-capturing lambda functions. For example, it cannot capture a local
```

> variable from the function. You can only use globally visible declarations.

Reviewer: ```suggestion
variable from the function, only globally visible declarations.
```

Reviewer: perhaps a bit more concise this way

> It's vital to ensure that the function does not throw any exceptions. Throwing exceptions from a `staticCFunction {}`

Reviewer: ```suggestion
Ensure that the function doesn't throw any exceptions. Throwing exceptions from a `staticCFunction {}`
```

> will end up in non-deterministic side effects.

Reviewer: (suggestion)
to get rid of the future-tense

## Use the C function pointer from Kotlin

> The next step is to call a C function pointer from a C pointer that you have from the `supply_fun()` call:

Reviewer: ```suggestion
The next step is to call the function pointer returned by `supply_fun()`:
```

Reviewer: not sure if this is accurate but the original is a bit awkward imho

Author: thank you! rephrased the whole sentence

Reviewer: looks good!

```kotlin
import interop.*
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.invoke

@OptIn(ExperimentalForeignApi::class)
fun myFun2() {
    val functionFromC = supply_fun() ?: error("No function is returned")

    functionFromC(42)
}
```

Kotlin turns the function pointer return type into a nullable `CPointer<CFunction<>` object. You need to first explicitly

> check for `null`; that's why the [Elvis operator](null-safety.md) is used in the code above.

Reviewer: (suggestion)
just to replace the ;

> The `cinterop` tool helps you turn a C function pointer into a simple `functionFromC(42)` object call in Kotlin.

Reviewer: instead of object call could we say function call on a function pointer object? (or rephrase this a bit?)

Something like:
The cinterop tool allows you to call a C function pointer as a regular Kotlin function, like `functionFromC(42)`
